src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs

```diff
@@ -420,6 +420,17 @@ internal List<CompletionResult> GetResultHelper(CompletionContext completionCont
 
                     case TokenKind.StringExpandable:
                     case TokenKind.StringLiteral:
+                        if (lastAst.Parent is CommandExpressionAst && lastAst.Parent.Parent is AssignmentStatementAst assignmentAst && assignmentAst.Left is VariableExpressionAst variableAst)
+                        {
+                            var value = completionContext.ExecutionContext.GetVariableValue(variableAst.VariablePath);
```

Comment:
This means that just having a variable that's already an enum will make PowerShell provide only that enum's members, right?

That would conflict in my mind with the idea of PowerShell being a dynamically typed language. This is a perfectly valid PowerShell program:

```powershell
enum Bird
{
    Duck
    Goose
}

$bird = [Bird]::Duck

$bird = 'Swan' # Valid

Write-Host "$bird"
```

I think, rather than provide values that fall into the same bucket as the current type, completions should offer possibilities that *won't fail*, which in this case depend on the *variable* rather than on the object that it currently refers to. You can constrain the possible values of a variable with a type constraint (or with an attribute more generally, but those are often less generative):

```powershell
enum Bird { Duck; Goose }

[Bird]$bird = 'Duck'

$bird = 'Swan' # Fails
```

To make this work, I'd do something like:

```csharp
// Pulls the variable out of an assignment's LHS expression
// Also brings back the innermost type constraint if there is one
private static VariableExpressionAst GetVariableFromExpressionAst(
    ExpressionAst expression,
    ref ITypeName typeConstraint,
    ref ValidateSetAttribute setConstraint)
{
    typeConstraint = null;
    switch (expression)
    {
        // $x = ...
        case VariableExpressionAst variableExpression:
            return variableExpression;

        // [type]$x = ...
        case ConvertExpressionAst convertExpression:
            typeConstraint = convertExpression.Type.TypeName;
            return GetVariableFromExpression(convertExpression.Child, ref typeConstraint, ref setConstraint);

        // [attribute()][type]$x = ...
        case AttributedExpressionAst attributedExpressionAst:

            try
            {
                setConstraint = attributedExpressionAst.Attribute.GetAttribute() as ValidateSetAttribute;
            }
            catch
            {
                // Do nothing, just prevent fallout from an unsuccessful attribute conversion
            }

            return GetVariableFromExpression(attributedExpressionAst.Child, ref typeConstraint, ref setConstraint);

        // Something else. Shouldn't be possible, but defend anyway
        default:
            return null;
    }
}

// Gets any type constraints or validateset constraints on a given variable
private bool TryGetTypeConstraintOnVariable(
    CompletionContext completionContext,
    string variableName,
    out Type typeConstraint,
    out ValidateSetAttribute setConstraint)
{
    typeConstraint = null;
    setConstraint = null;

    PSVariable variable = completionContext.ExecutionContext.EngineSessionState.GetVariable(variableName);

    if (variable?.Attributes == null || variable.Attributes.Count == 0)
    {
        return null;
    }

    foreach (Attribute attribute in variable.Attributes)
    {
        if (attribute is ArgumentTypeConverterAttribute typeConverterAttribute)
        {
            typeConstraint = typeConverterAttribute.TargetType;
            continue;
        }

        if (attribute is ValidateSetAttribute validateSetAttribute)
        {
            setConstraint = validateSetAttribute;
        }
    }

    return typeConstraint != null || setConstraint != null;
}

private bool TryGetCompletionsForVariableAssignment(
    CompletionContext completionContext,
    AssignmentStatementAst assignmentAst,
    out IReadOnlyList<CompletionResult> completions)
{
    // Try to get the variable from the assignment, plus any type constraint on it
    VariableExpressionAst variableAst = GetVariableFromExpression(assignmentAst.Left, ref ITypeName typeConstraint, ref ValidateSetAttribute setConstraint);

    if (variableAst == null)
    {
        completions = null;
        return false;
    }

    // Assignment constraints override any existing ones, so try them first

    // Check any [ValidateSet()] constraint first since it's likely to be narrow
    if (setConstraint?.ValidValues != null)
    {
        completions = GetResultForSet(setConstraint.ValidValues, completionContext);
        return true;
    }

    // Then try to complete for an enum type
    Type variableType = typeConstraint?.GetReflectionType();
    if (variableType != null)
    {
        if (!variableType.IsEnum)
        {
            completions = null;
            return false;
        }

        completions = GetResultForEnum(variableType, completionContext);
        return true;
    }

    // If the assignment itself was unconstrained, the variable still might be
    if (!TryGetTypeConstraintOnVariable(completionContext, variableAst.VariablePath, out typeConstraint, out setConstraint))
    {
        completions = null;
        return false;
    }

    // Again try the [ValidateSet()] constraint first
    if (setConstraint?.ValidValues != null)
    {
        completions = GetResultForSet(setConstraint.ValidValues, completionContext);
        return true;
    }

    // Then try to complete for an enum type
    Type variableType = typeConstraint?.GetReflectionType();
    if (variableType != null)
    {
        if (!variableType.IsEnum)
        {
            completions = null;
            return false;
        }

        completions = GetResultForEnum(variableType, completionContext);
        return true;
    }
}

...

// Search to see if we're looking at an assignment
if (lastAst.Parent is CommandExpressionAst
    && lastAst.Parent.Parent is AssignmentStatementAst assignmentAst)
{
    if (TryGetCompletionsForVariableAssignment(completionContext, assignmentAst, out IReadOnlyList<CompletionResult> completions))
    {
        return completions;
    }
}
```

Btw I haven't tried to compile or run this code, it's just me trying to get my thoughts down.

Reply:
Wait... Based on _what_ exactly are you suggesting to complete arbitrary (undefined?) strings? Paths? Something else? What additional completions are sensible for a variable like this? Currently I didn't think we offered any explicit completions for this kind of case?

Reply:
Yeah there are no completions for this case currently.

The original request was to provide completions for preference variables:

```powershell
$ErrorActionPreference = <TAB>
```

The current implementation here looks up the variable's type and *if it's an enum* it provides the possible enum members as completions.

My submission is that it's not the variable's value's type, but the type/set constraints on the variable. So the code I've got above purports to implement that. Haven't tested it of course...

Reply:
> Based on what exactly are you suggesting to complete arbitrary (undefined?) strings? Paths? Something else? 

Only for *variable assignment*:

```
$x = <TAB>
```

Reply:
Gotcha!

Sorry, I got the wrong impression of what you said originally. That sounds sensible enough. 🙂

Reply:
@rjmholt I think this is a good suggestion but beyond the scope of the intent of this PR.  I created https://github.com/PowerShell/PowerShell/issues/10694 to track this.

Reply:
will update